### ISSUE TITLE
[Bank Account] Threading Tests Incorrectly Passing on Website, Correctly Failing Locally

### ISSUE THREAD

**Issue by mohmad-null** [exercism/python#2821]
The bank account problem is meant to be "hard" with the difficulty coming from the need to lock the account during threading.

I created a solution that passed all of the other tests and completely ignored the threading part (there was no code at all for that). On my local PC it passed all tests except the threading one, exactly as expected.

Except when I copy/pasted it to exercism and ran the tests there, *all* of the tests passed and I could submit my solution.

(Edit: Also the Intro blurb for Bank Account says the solution needs to handle "threading/processes" depending on what the language is. Given Python supports both of these but the code is only using Threading, this should probably be changed to just threading).

**Comment by BethanyG** [comment]
Hi @mohmad-null,

Thanks for logging this issue!  
The `TL;DR` is that this works "as expected" on a local environment, but does **not** behave "as expected" within Docker (_which is how our test runner works in production_).  Current working suspicion is that Docker is limiting the number of threads, thereby short circuiting the behavior needed to create a locking issue.

This one is tricky, and I have been thinking of soliciting help from folx with more experience in threading/processing within Docker to rework it.  Here are some details from a Slack discussion around this from about three months ago:  https://exercism-team.slack.com/archives/CC5DGSAG6/p1633825675090000.  Screengrab below, if you cannot access Slack:

![2021-12-19_13-10-46](https://user-images.githubusercontent.com/5923094/146690959-9ede07ca-aeee-42dc-b6f2-d03dbd4f1675.png)

Happy to hear your thoughts on how to go about fixing this one.

As for the directions -- those are pulled from [problem-specifications](https://github.com/exercism/problem-specifications/blob/main/exercises/bank-account/description.md), so changes would need to be proposed there, or we would need to work out Python-specific language for an addendum that pointed students toward using Threading.

**Comment by mohmad-null** [comment]
I had a friend complete the Bank Account one, and like mine, they made no attempt to do the locking thing. Their solution was broadly similar to mine.

For them, *all* tests passed in their local environment(!), but the concurrency test correctly failed when the code was copy/pasted to exercism. So the exact opposite behaviour to what I was getting.

**Comment by BethanyG** [comment]
We found similar issues when we were testing/investigating.  I could get a pass or a fail on my Mac by reducing thread count as noted in the Slack thread -- but only under some other OS load.  Pass/Fail wasn't exactly reliable.  @J08K had more reliable luck using Windows.  The website seemed more reliably bad/incorrect -- although 1 out of 4 would fail "as expected", with the other three passing.

So definitely a case for someone enthusiastic about multithreading in Python to dig into.  😄   This exercise might need a bit of a re-think testing-wise.  LMK if you'd like to take it on.  😉

**Comment by BethanyG** [comment]
@IsaacG @mohmad-null -

I am going to close this.  We recently merged [canonical data](https://github.com/exercism/problem-specifications/pull/2192) for this exercise in problem-specifications, and as part of the maintenance work, we created a JinJa2 template, a new example, and a `tests.toml` file.  Rather than fiddle around to see if I could create an effective concurrency test against the now much faster Python 3.11, I instead commented out the concurrency test case.

If either of you would like to take on making an effective case, LMK and I will re-open.
